### PR TITLE
Feature/add edit event user endpoint

### PR DESCRIPTION
**Pull request by liza459**
https://trello.com/c/8bGTGxEN/220-eventservice-add-eventuser-edit-endpoint

https://docs.google.com/document/d/1upUYu5MaGjBW2MKP-wHM2hwef8t6TPGzpVNkF8njJDI/edit#heading=h.7jjwan5qecem

Also
- rights check fix in CreateEventUserCommand;
- fix EventUserInfoMapper;

### CODE REVIEW

File: Common.Configurations.targets
Comment (SolovyovDmitry) [inline Common.Configurations.targets:1]: Why did you change spacing? And we have those spaces in the end everywhere

File: src/EventService.Broker/EventService.Broker.csproj
Comment (SolovyovDmitry) [inline src/EventService.Broker/EventService.Broker.csproj:1]: It shouldn't be here

File: src/EventService.Business/Commands/EventUser/CreateEventUserCommand.cs [hunk 1]
```diff
@@ -86,12 +87,9 @@ public async Task<OperationResultResponse<bool>> ExecuteAsync(CreateEventUserReq
...
     }
 
-    if ((dbEvent.Access == AccessType.Closed && !await _accessValidator.HasRightsAsync(senderId, Rights.AddEditRemoveUsers))
-        || !(dbEvent.Access == AccessType.Opened
-          && (await _accessValidator.HasRightsAsync(senderId, Rights.AddEditRemoveUsers)
-            || (!await _accessValidator.HasRightsAsync(senderId, Rights.AddEditRemoveUsers)
-              && request.Users.Count == 1
-              && request.Users.Exists(x => x.UserId == senderId)))))
+    if ((dbEvent.Access == AccessType.Closed && !userWithRight) ||
+      (dbEvent.Access == AccessType.Opened &&
```
Comment (SolovyovDmitry) [inline src/EventService.Business/Commands/EventUser/CreateEventUserCommand.cs]: format it to be at the same level

File: src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      (await _eventUserRepository.GetAsync(eventUserId)).UserId != senderId)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.Forbidden);
+    }
+
+    ValidationResult validationResult = await _validator.ValidateAsync((eventUserId, patch));
+    if (!validationResult.IsValid)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.BadRequest,
+        validationResult.Errors.Select(vf => vf.ErrorMessage).ToList());
```
Comment (SolovyovDmitry) [inline src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs]: Select().ToList() => ConvertAll()

File: src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    _responseCreator = responseCreator;
+    _accessValidator = accessValidator;
+    _httpContextAccessor = httpContextAccessor;
+    _contextAccessor = contextAccessor;
+  }
+  public async Task<OperationResultResponse<bool>> ExecuteAsync(Guid eventUserId, JsonPatchDocument<EditEventUserRequest> patch)
+  {
+    Guid senderId = _httpContextAccessor.HttpContext.GetUserId();
+
+    if (!await _accessValidator.HasRightsAsync(Rights.AddEditRemoveUsers) &&
```
Comment (SolovyovDmitry) [inline src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs:54]: format

File: src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    ValidationResult validationResult = await _validator.ValidateAsync((eventUserId, patch));
+    if (!validationResult.IsValid)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.BadRequest,
+        validationResult.Errors.Select(vf => vf.ErrorMessage).ToList());
+    }
+
+    OperationResultResponse<bool> response = new();
```
Comment (SolovyovDmitry) [inline src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs]: Initialization can be simplified

File: src/EventService.Business/EventService.Business.csproj
Comment (SolovyovDmitry) [inline src/EventService.Business/EventService.Business.csproj:1]: Shouldn't be here

File: src/EventService.Data.Provider.MsSql.Ef/EventService.Data.Provider.MsSql.Ef.csproj
Comment (SolovyovDmitry) [inline src/EventService.Data.Provider.MsSql.Ef/EventService.Data.Provider.MsSql.Ef.csproj:1]: same

File: src/EventService.Data.Provider/EventService.Data.Provider.csproj
Comment (SolovyovDmitry) [inline src/EventService.Data.Provider/EventService.Data.Provider.csproj:1]: same

File: src/EventService.Data/EventService.Data.csproj
Comment (SolovyovDmitry) [inline src/EventService.Data/EventService.Data.csproj:1]: same

File: src/EventService.Data/EventUserRepository.cs [hunk 5]
```diff
@@ -53,4 +60,27 @@ public async Task<bool> CreateAsync(List<DbEventUser> dbEventUsers)
 
     return eventUsersQuery.ToListAsync(cancellationToken: cancellationToken);
   }
+
+  public Task<DbEventUser> GetAsync(Guid eventUserId)
+  {
+    return _provider.EventsUsers.AsNoTracking().FirstOrDefaultAsync(eu =>eu.Id == eventUserId);
```
Comment (SolovyovDmitry) [inline src/EventService.Data/EventUserRepository.cs]: missing space before eu.Id

File: src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  private readonly IAccessValidator _accessValidator;
+  private readonly IHttpContextAccessor _httpContextAccessor;
+
+  private void HandleInternalPropertyValidationAsync(
+    Operation<EditEventUserRequest> requestedOperation,
+    DbEvent dbEvent,
+    DbEventUser dbEventUser,
+    bool isAddEditRemoveUsers,
+    bool isUser,
+    ValidationContext<(Guid,JsonPatchDocument<EditEventUserRequest>)> context)
```
Comment (SolovyovDmitry) [inline src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs]: missing space

File: src/EventService/Controllers/EventUserController.cs [hunk 7]
```diff
@@ -31,4 +33,13 @@ public class EventUserController : ControllerBase
...
     return await command.ExecuteAsync(eventId: eventId, filter: filter, cancellationToken: cancellationToken);
   }
+
+  [HttpPatch("edit")]
+    public async Task<OperationResultResponse<bool>> EditAsync(
+      [FromServices] IEditEventUserCommand command,
+      [FromQuery] Guid eventUserId,
+      [FromBody] JsonPatchDocument<EditEventUserRequest> request)
+  {
+    return await command.ExecuteAsync(eventUserId,request);
```
Comment (SolovyovDmitry) [inline src/EventService/Controllers/EventUserController.cs]: missing space

**Review by SolovyovDmitry — CHANGES_REQUESTED** [review]
I'd suggest running code cleanup. Default profile made 20 files change with missing spaces, unused usings etc. Remove updates from .csproj and insert the block we always use. The logic itself looks OK, but I will check harder a bit later

File: Common.Configurations.targets
Comment (liza459) [reply Common.Configurations.targets:1]: изображение

There are no spaces at the beginning of the line, this is a git bug.

File: src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+          x => Enum.TryParse(typeof(EventUserStatus), x.value?.ToString(), out _) ?
+            (x.value.ToString().Trim() == "Participant" &&
+                (((dbEventUser.Status == EventUserStatus.Refused || dbEventUser.Status == EventUserStatus.Invited) && isUser)||
+                  (dbEventUser.Status == EventUserStatus.Discarded && isAddEditRemoveUsers)))||
+              (x.value.ToString().Trim() == "Refused" &&
+                  (dbEventUser.Status == EventUserStatus.Participant || dbEventUser.Status == EventUserStatus.Invited) && isUser)||
+                (x.value.ToString().Trim() == "Discarded" &&
+                  dbEventUser.Status == EventUserStatus.Participant && isAddEditRemoveUsers)
+            : false,
+            "Uncorrect user status"
```
Comment (Draktolli) [inline src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs]: In my Opinion it's better to use 
`Enum.TryParse(x.value?.ToString(), out EventUserStatus status) ?
status == EventUserStatus.Participant &&`
and etc

Comment (Draktolli) [inline src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs]: In other projects messages was writen in .resx file

File: Common.Configurations.targets
Comment (TIove) [reply Common.Configurations.targets:1]: That's not a bug, there were **'spaces'** and became **'tabs'**. In VS tabs look like spaces, but it doesn't. 

@SolovyovDmitry What do u mean about **'those spaces in the end'**? Does Liza have to delete them or not?

File: src/EventService.Business/Commands/EventUser/CreateEventUserCommand.cs [hunk 9]
```diff
@@ -78,6 +78,7 @@ public async Task<OperationResultResponse<bool>> ExecuteAsync(CreateEventUserReq
   {
     Guid senderId = _contextAccessor.HttpContext.GetUserId();
     DbEvent dbEvent = await _eventRepository.GetAsync(request.EventId);
+    bool userWithRight = await _accessValidator.HasRightsAsync(senderId, Rights.AddEditRemoveUsers);
```
Comment (TIove) [inline src/EventService.Business/Commands/EventUser/CreateEventUserCommand.cs]: userHasRight is prettier, fmpov

File: src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  {
+    _validator = validator;
+    _eventUserRepository = eventUserRepository;
+    _mapper = mapper;
+    _responseCreator = responseCreator;
+    _accessValidator = accessValidator;
+    _httpContextAccessor = httpContextAccessor;
+    _contextAccessor = contextAccessor;
+  }
+  public async Task<OperationResultResponse<bool>> ExecuteAsync(Guid eventUserId, JsonPatchDocument<EditEventUserRequest> patch)
```
Comment (TIove) [inline src/EventService.Business/Commands/EventUser/EditEventUserCommand.cs]: Space after ctor. And separate params to few lines.

File: src/EventService.Models.Dto/Requests/EventUser/EditEventUserRequest.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using LT.DigitalOffice.EventService.Models.Dto.Enums;
+
+namespace LT.DigitalOffice.EventService.Models.Dto.Requests.EventUser
```
Comment (TIove) [inline src/EventService.Models.Dto/Requests/EventUser/EditEventUserRequest.cs]: namespace without brackets

File: src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs (hunk 8 above)
Comment (TIove) [inline src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs]: Incorrect*

File: src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+      nameof(EditEventUserRequest.NotifyAtUtc),
+      x => x == OperationType.Replace,
+      new Dictionary<Func<Operation<EditEventUserRequest>, bool>, string>
+      {
+        {
+          x => string.IsNullOrEmpty(x.value?.ToString())? true :
+            (DateTime.TryParse(x.value.ToString().Trim(), out DateTime date) &&
+            DateTime.Parse(x.value.ToString().Trim()) < dbEvent.Date &&
+            DateTime.Parse(x.value.ToString().Trim()) > DateTime.UtcNow),
+            "Uncorrect notify"
```
Comment (TIove) [inline src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs]: Incorrect*

File: Common.Configurations.targets
Comment (SolovyovDmitry) [reply Common.Configurations.targets:1]: @TIove I mean these
image

Comment (SolovyovDmitry) [reply Common.Configurations.targets:1]: in the end of the lines

File: src/EventService.Data/EventUserRepository.cs [hunk 13]
```diff
@@ -48,9 +54,32 @@ public async Task<bool> CreateAsync(List<DbEventUser> dbEventUsers)
 
     if (filter.Status is not null)
```
Comment (SolovyovDmitry) [inline src/EventService.Data/EventUserRepository.cs]: can it be !HasValue?

File: src/EventService.Mappers/Models/EventUserInfoMapper.cs [hunk 14]
```diff
@@ -13,9 +13,9 @@ public List<EventUserInfo> Map(List<UserInfo> userInfos, List<DbEventUser> event
     return eventUsers?.Select(eu => new EventUserInfo
     {
       Id = eu.Id,
-      Status = eu.Status,
+      Status = eu.Status.ToString(),
```
Comment (SolovyovDmitry) [inline src/EventService.Mappers/Models/EventUserInfoMapper.cs]: Why did you decide to change?

**Review by SolovyovDmitry — APPROVED** [review]

**Review by SolovyovDmitry — APPROVED** [review]

File: src/EventService.Mappers/Models/EventUserInfoMapper.cs (hunk 14 above)
Comment (liza459) [reply src/EventService.Mappers/Models/EventUserInfoMapper.cs]: I tried to fix the fact that the enum is returned as a number. It was a bad decision.

Now added an attribute ( `[JsonConverter(typeof(StringEnumConverter))]`) to the model.

File: src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+      new Dictionary<Func<Operation<EditEventUserRequest>, bool>, string>
+      {
+        {
+          x => Enum.TryParse(typeof(EventUserStatus), x.value?.ToString(), out _),
+              EventUserRequestValidatorResource.IncorrectFormatStatus
+        },
+        {
+          x => Enum.TryParse(x.value?.ToString(), out EventUserStatus newStatus) &&
+              ((newStatus == EventUserStatus.Participant &&
+                (((status == EventUserStatus.Refused || status == EventUserStatus.Invited) && isUser)||
```
Comment (VLiamin) [inline src/EventService.Validation/EventUser/EditEventUserRequestValidator.cs]: May be add space before |

**Review by VLiamin — APPROVED** [review]

**Review by TIove — APPROVED** [review]

**Review by TheDioBrando — APPROVED** [review]

**Review by Segway-svg — APPROVED** [review]

**Pull request merged**